### PR TITLE
Add mask_soft config option to allow non-binary masks

### LINKED ISSUE #4606: Soft Segmentation Masks
Currently in the segmentation heads it appears that masks are always given as binary targets.

For some problems, it may be more natural to instead express the ground truth as either a weighted mask, or more simply a mask of fractional "soft" values.

From what I can tell, this might be as simple as conditionally executing `resized_masks = (targets >= 0.5).cpu().numpy()` in the `crop_and_resize` function of `BitmapMasks` (in mmdet/core/mask/structures.py) based on the `cfg` passed to `mask_target` (in mmdet/core/mask/mask_target.py).

### PR DESCRIPTION
**Pull request by Erotemic**
Closes #4606

When using mmdetection as a library for MaskRCNN there is almost nothing preventing the ground truth masks from being fractional labels instead of hard binary labels. The only issue seemed to be in `BitmapMasks.crop_and_resize` where after a resize the interpolated values were always rounded to 0 or 1.

To prevent this I simply added a keyword argument `rounded=True`, which if set to `False` will prevent this rounding step. I also added a small bit of logic to `mask_target` such that it looks for the configuration option `mask_soft` (defaulted to `False`), and it sets `rounded = not mask_soft`.

In the case for mask Polygons I do expose the `rounded` parameter for API consistency, but I raise a ValueError if it is false because soft masks don't make much sense for Polygon targets.

Thus I was able to simply add `roi_head.train_cfg['mask_soft'] = True` to my MaskRCNN training configuration and the mask loss seems to be computed correctly against these soft targets.

Again, I'm using mmdetection as a library to define models (i.e. I'm not using my own training loop and data loading scripts), so I don't think this feature would be exposed to mmdet script users without further changes, but this should be the core fundamental change necessary to add soft mask targets to MaskHeads.

### CODE REVIEW

File: mmdet/core/mask/structures.py [hunk 1]
```diff
@@ -323,7 +323,7 @@ def crop_and_resize(self,
                         out_shape,
                         inds,
                         device='cpu',
-                        interpolation='bilinear'):
+                        interpolation='bilinear', rounded=True):
```
Comment (hellock) [inline mmdet/core/mask/structures.py:326]: rounded -> binarize

Comment (hellock) [reply mmdet/core/mask/structures.py:326]: The docstring of the base class needs to be updated.

**Comment by hellock** [comment]
Thanks for this nice feature. Could you also fix the linting errors?

**Comment by ZwwWayne** [comment]
Kindly ping @Erotemic

**Comment by ZwwWayne** [comment]
The PR looks good to me now. We will benchmark it in May and will merge it for v2.13.0

**Review by ZwwWayne — APPROVED** [review]

**Comment by RangiLyu** [comment]
I have benchmarked mask_rcnn_r50_fpn with `soft_mask_target=True` on COCO. 
The bbox eval results are
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.382
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=1000 ] = 0.590
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=1000 ] = 0.416
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.223
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.417
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.495
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.522
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=300 ] = 0.522
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=1000 ] = 0.522
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.326
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.566
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.660
```
And the segm eval results are

```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.345
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=1000 ] = 0.556
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=1000 ] = 0.368
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.185
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.377
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.469
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.474
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=300 ] = 0.474
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=1000 ] = 0.474
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.276
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.516
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.623
```

**Comment by Erotemic** [comment]
@RangiLyu Can you link to the comparable results where `soft_mask_target=False` so it is clear if there is any difference?

**Comment by Erotemic** [comment]
Thanks @RangiLyu for pulling up these numbers. The difference is likely due to non-determinism due to cudnn, parallelism, data loading, etc. The only place where the objective is changing is on boundaries in the mask truth (there will be a few fractional values along the boundary due to interpolation).  I imagine if you ran both experiments multiple times, you would find the distribution of APs would be difficult to separate (i.e. p > 0.5). But then again, perhaps fuzzy edges do cause a real 0.2 drop in the score.

But it seems likely there are no significant regressions, so this seems ready for a merge.

**Comment by ZwwWayne** [comment]
Thanks @Erotemic & @RangiLyu
